src/MongoDB.Bson/IO/JsonReader.cs

```diff
@@ -1873,14 +1876,26 @@ private BsonValue ParseRegularExpressionConstructor()
             return new BsonRegularExpression(patternToken.StringValue, options);
         }
 
-        private BsonValue ParseRegularExpressionExtendedJsonLegacy()
+        private bool TryParseRegularExpressionExtendedJsonLegacy(out BsonValue value)
         {
             VerifyToken(":");
             var patternToken = PopToken();
             if (patternToken.Type != JsonTokenType.String)
             {
-                var message = string.Format("JSON reader expected a string but found '{0}'.", patternToken.Lexeme);
-                throw new FormatException(message);
+                PushToken(patternToken);
+                VerifyToken("{");
+                var nameToken = PopToken();
+                if (nameToken.Type != JsonTokenType.String)
+                {
+                    throw new FormatException($"JSON reader expected a string but found '{nameToken.Lexeme}'.");
+                }
+                if (nameToken.StringValue != "$regularExpression")
+                {
+                    throw new FormatException($"JSON reader expected a \"$regularExpression\" but found '{nameToken.StringValue}'.");
+                }
+                value = null;
+
+                return false;
```

Comment:
I think this change is too extensive. Instead of changing lines 1882-1883 (left side) just add this after line 1882 (right side):

    if (patternToken.Type == JsonTokenType.BeginObject)
    {
        value = null;
        return false;
    }

Reply:
Done.
Note: I restored the previous error message as well.